File: Julia_research/QuantEngine/src/production/ntp_check.jl
# ── NTP Check — Clock Skew Detection ──────────────────────────────────
# Trading systems need accurate timestamps.
# Even 300-800ms skew causes cancel races on Binance futures.
#
# Thresholds (per feedback):
#   < 50ms:   healthy
#   50-150ms: warning
#   > 150ms:  critical — should block trading
#
# Tightened from 1000ms to 150ms based on production experience.

using Sockets
using Dates

const NTP_SERVERS = ["time.apple.com", "time.google.com", "pool.ntp.org"]
const NTP_EPOCH_OFFSET = 2208988800  # seconds between 1900-01-01 and 1970-01-01

struct NTPResult
    skew_ms::Float64
    server::String
    level::Symbol       # :healthy, :warning, :critical
    checked_at::DateTime
end

"""
Query an NTP server and return clock skew in seconds.
Uses simplified SNTP (UDP port 123).
"""
function query_ntp(server::String; timeout_s::Float64=2.0)::Float64
    sock = UDPSocket()

    try
        # Resolve hostname
        addr = getaddrinfo(server)

        # Build NTP request packet (48 bytes, LI=0, VN=3, Mode=3)
        packet = zeros(UInt8, 48)
        packet[1] = 0x1b  # LI=0, VN=3, Mode=3 (client)

        t1 = time()
        send(sock, addr, 123, packet)

        # Wait for response with timeout
        response = Channel{Vector{UInt8}}(1)
        @async begin
            try
                data, _ = recvfrom(sock)
                put!(response, data)
            catch
            end
        end

        result = timedwait(() -> isready(response), timeout_s)
        if result === :timed_out
            error("NTP timeout from $server")
        end

        t4 = time()
        data = take!(response)

        # Parse transmit timestamp (bytes 41-44 = seconds since 1900)
        transmit_seconds = UInt32(data[41]) << 24 | UInt32(data[42]) << 16 |
                          UInt32(data[43]) << 8  | UInt32(data[44])
        t3 = Float64(transmit_seconds) - NTP_EPOCH_OFFSET

        # Simplified offset
        rtt = t4 - t1
        offset = t3 - t1 - rtt / 2

        return offset
    finally
        close(sock)
    end
end

"""
Check clock skew against NTP servers.
Returns NTPResult with skew and severity level.
Threshold: 150ms critical (tightened per Binance cancel race feedback).
"""
function check_ntp(;
    warning_ms::Float64 = 50.0,
    critical_ms::Float64 = 150.0,
    servers::Vector{String} = NTP_SERVERS,
)::NTPResult
    for server in servers
        try
            offset = query_ntp(server)
            skew_ms = abs(offset) * 1000.0

            level = if skew_ms < warning_ms
                :healthy
            elseif skew_ms < critical_ms
                @warn "Clock skew: $(round(skew_ms, digits=1))ms (server=$server)"
                :warning
            else
                @error "CRITICAL clock skew: $(round(skew_ms, digits=1))ms — cancel races possible"
                :critical
            end

            return NTPResult(skew_ms, server, level, now(UTC))
        catch e
            @debug "NTP query failed for $server: $e"
            continue
        end
    end

    @warn "All NTP servers unreachable"
    return NTPResult(0.0, "unreachable", :unknown, now(UTC))
end

"""Should trading be blocked due to clock skew?"""
function ntp_blocks_trading(result::NTPResult)::Bool
    return result.level == :critical
end

File: TradingSwarm/src/monitoring/ntp_check.py
"""
NTP Validation + Clock Skew Detection.

Trading systems require accurate timestamps. This module:
- Checks local clock against NTP servers
- Detects drift above configurable thresholds
- Blocks trading on excessive skew
- Logs skew as a health metric
"""

from __future__ import annotations

import asyncio
import logging
import struct
import time
from typing import Any

logger = logging.getLogger(__name__)

NTP_SERVERS = [
    "time.apple.com",
    "time.google.com",
    "pool.ntp.org",
]

# NTP epoch is 1900-01-01, Unix epoch is 1970-01-01
NTP_DELTA = 2208988800


class NTPCheck:
    """
    Validates local clock against NTP servers.

    Thresholds:
    - < 100ms: healthy
    - 100ms - 1s: warning
    - > 1s: critical (should block trading)
    """

    def __init__(
        self,
        warning_ms: float = 100.0,
        critical_ms: float = 1000.0,
    ) -> None:
        self._warning_ms = warning_ms
        self._critical_ms = critical_ms
        self._last_skew_ms: float | None = None
        self._last_check: float = 0

    @property
    def is_healthy(self) -> bool:
        if self._last_skew_ms is None:
            return True  # unknown = assume OK
        return abs(self._last_skew_ms) < self._critical_ms

    @property
    def last_skew_ms(self) -> float | None:
        return self._last_skew_ms

    async def check(self, server: str | None = None) -> dict[str, Any]:
        """
        Query NTP and return skew info.

        Returns dict with: skew_ms, server, healthy, level
        """
        servers = [server] if server else NTP_SERVERS

        for srv in servers:
            try:
                skew = await self._query_ntp(srv)
                self._last_skew_ms = skew * 1000
                self._last_check = time.time()

                abs_ms = abs(self._last_skew_ms)
                if abs_ms < self._warning_ms:
                    level = "healthy"
                elif abs_ms < self._critical_ms:
                    level = "warning"
                    logger.warning("Clock skew: %.1fms (server=%s)", self._last_skew_ms, srv)
                else:
                    level = "critical"
                    logger.critical(
                        "CRITICAL clock skew: %.1fms (server=%s) — trading should be blocked",
                        self._last_skew_ms, srv,
                    )

                return {
                    "skew_ms": self._last_skew_ms,
                    "server": srv,
                    "healthy": level != "critical",
                    "level": level,
                }
            except Exception:
                logger.debug("NTP query failed for %s", srv)
                continue

        logger.warning("All NTP servers unreachable")
        return {"skew_ms": None, "server": None, "healthy": True, "level": "unknown"}

    async def _query_ntp(self, server: str) -> float:
        """
        Query a single NTP server. Returns skew in seconds.

        Uses raw UDP socket with SNTP (simplified NTP).
        """
        loop = asyncio.get_event_loop()

        def _sync_query() -> float:
            import socket
            client = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            client.settimeout(2)

            # NTP request packet
            data = b'\x1b' + 47 * b'\0'
            t1 = time.time()
            client.sendto(data, (server, 123))
            response, _ = client.recvfrom(1024)
            t4 = time.time()
            client.close()

            # Parse transmit timestamp (bytes 40-47)
            transmit_time = struct.unpack('!12I', response)[10]
            t3 = transmit_time - NTP_DELTA

            # Simplified offset: server_time - local_time
            rtt = t4 - t1
            offset = t3 - t1 - rtt / 2
            return offset

        return await loop.run_in_executor(None, _sync_query)

    async def periodic_check(self, interval_seconds: float = 300.0) -> None:
        """Run NTP check periodically."""
        while True:
            result = await self.check()
            if result["level"] == "critical":
                from src.monitoring.alerting import AlertSeverity, Alert, get_alert_manager
                mgr = get_alert_manager()
                await mgr.send(Alert(
                    AlertSeverity.CRITICAL,
                    "Clock Skew Critical",
                    f"Skew: {result['skew_ms']:.1f}ms — trading unsafe",
                    source="ntp_check",
                ))
            await asyncio.sleep(interval_seconds)
